### PR TITLE
Harden ResponseCache (and fix #9507)

### LINKED ISSUE #9507: Synapse potentially cached a failed sync request
Running https://github.com/matrix-org/synapse/pull/9379/commits/4a3260092d4c009cdd0c274a83adb89bbef72c83

Last night my database suffered an issue which meant that it was briefly unavailable, and it dropped all connections for a period. Synapse *seemed* to largely recover as sync requests to some accounts are working, and federation/event creation seems to be working fine. 

```
2021-02-27 13:10:17,830 - synapse.util.caches.response_cache - 191 - INFO - GET-84929- [sync]: using incomplete cached result for [(@Half-Shot:half-shot.uk, 0, 's44209228_293425298_29572_27619593_543941_1697_969558_19742008_52', '183', False, 'HEWQINNMGI')]
2021-02-27 13:16:43,781 - synapse.access.http.8018 - 299 - INFO - GET-84929- 86.16.94.189 - 8018 - {@Half-Shot:half-shot.uk} Processed request: 384.703sec/1.250sec (0.252sec, 0.000sec) (0.000sec/0.000sec/0) 1877428B 200 "GET /_matrix/client/r0/sync?filter=183&timeout=0&since=s44209228_293425298_29572_27619593_543941_1697_969558_19742008_52 HTTP/1.0" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Element/1.7.21 Chrome/85.0.4183.121 Electron/10.2.0 Safari/537.36" [0 dbevts]
```

(ignore the absolutely awful sync duration, it's sadly a new normal atm)

However, sync requests for my Element android device fail with a 500 each time:

```
2021-02-27 13:18:55,369 - synapse.util.caches.response_cache - 189 - INFO - GET-85063- [sync]: using completed cached result for [(@Half-Shot:half-shot.uk, 0, None, '181', False, 'DHNXFYGPRT')]
2021-02-27 13:18:55,369 - twisted - 257 - CRITICAL - GET-85063- Unhandled error in Deferred:
2021-02-27 13:18:55,370 - twisted - 257 - CRITICAL - GET-85063- 
Traceback (most recent call last):
  File "/opt/synapse/lib/python3.9/site-packages/synapse/storage/database.py", line 532, in new_transaction
    r = func(cursor, *args, **kwargs)
  File "/opt/synapse/lib/python3.9/site-packages/synapse/storage/databases/state/bg_updates.py", line 121, in _get_state_groups_from_groups_txn
    txn.execute(sql % (where_clause,), args)
  File "/opt/synapse/lib/python3.9/site-packages/synapse/storage/database.py", line 294, in execute
    self._do_execute(self.txn.execute, sql, *args)
  File "/opt/synapse/lib/python3.9/site-packages/synapse/storage/database.py", line 320, in _do_execute
    return func(sql, *args)
psycopg2.OperationalError: SSL SYSCALL error: EOF detected


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/synapse/lib/python3.9/site-packages/twisted/internet/defer.py", line 1416, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/opt/synapse/lib/python3.9/site-packages/twisted/python/failure.py", line 512, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "/opt/synapse/lib/python3.9/site-packages/synapse/handlers/sync.py", line 313, in _wait_for_sync_for_user
    result = await self.current_sync_for_user(
  File "/opt/synapse/lib/python3.9/site-packages/synapse/handlers/sync.py", line 344, in current_sync_for_user
    return await self.generate_sync_result(sync_config, since_token, full_state)
  File "/opt/synapse/lib/python3.9/site-packages/synapse/handlers/sync.py", line 1000, in generate_sync_result
    res = await self._generate_sync_entry_for_rooms(
  File "/opt/synapse/lib/python3.9/site-packages/synapse/handlers/sync.py", line 1444, in _generate_sync_entry_for_rooms
    await concurrently_execute(handle_room_entries, room_entries, 10)
  File "/opt/synapse/lib/python3.9/site-packages/twisted/internet/defer.py", line 1416, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/opt/synapse/lib/python3.9/site-packages/twisted/python/failure.py", line 512, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "/opt/synapse/lib/python3.9/site-packages/synapse/util/async_helpers.py", line 174, in _concurrently_execute_inner
    await maybe_awaitable(func(next(it)))
  File "/opt/synapse/lib/python3.9/site-packages/synapse/handlers/sync.py", line 1432, in handle_room_entries
    res = await self._generate_room_entry(
  File "/opt/synapse/lib/python3.9/site-packages/synapse/handlers/sync.py", line 1829, in _generate_room_entry
    batch = await self._load_filtered_recents(
  File "/opt/synapse/lib/python3.9/site-packages/synapse/handlers/sync.py", line 516, in _load_filtered_recents
    loaded_recents = await filter_events_for_client(
  File "/opt/synapse/lib/python3.9/site-packages/synapse/visibility.py", line 84, in filter_events_for_client
    event_id_to_state = await storage.state.get_state_for_events(
  File "/opt/synapse/lib/python3.9/site-packages/synapse/storage/state.py", line 467, in get_state_for_events
    group_to_state = await self.stores.state._get_state_for_groups(
  File "/opt/synapse/lib/python3.9/site-packages/synapse/storage/databases/state/store.py", line 257, in _get_state_for_groups
    group_to_state_dict = await self._get_state_groups_from_groups(
  File "/opt/synapse/lib/python3.9/site-packages/synapse/storage/databases/state/store.py", line 162, in _get_state_groups_from_groups
    res = await self.db_pool.runInteraction(
  File "/opt/synapse/lib/python3.9/site-packages/synapse/storage/database.py", line 660, in runInteraction
    result = await self.runWithConnection(
  File "/opt/synapse/lib/python3.9/site-packages/synapse/storage/database.py", line 743, in runWithConnection
    return await make_deferred_yieldable(
  File "/opt/synapse/lib/python3.9/site-packages/twisted/python/threadpool.py", line 250, in inContext
    result = inContext.theWork()
  File "/opt/synapse/lib/python3.9/site-packages/twisted/python/threadpool.py", line 266, in <lambda>
    inContext.theWork = lambda: context.call(ctx, func, *args, **kw)
  File "/opt/synapse/lib/python3.9/site-packages/twisted/python/context.py", line 122, in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "/opt/synapse/lib/python3.9/site-packages/twisted/python/context.py", line 85, in callWithContext
    return func(*args,**kw)
  File "/opt/synapse/lib/python3.9/site-packages/twisted/enterprise/adbapi.py", line 306, in _runWithConnection
    compat.reraise(excValue, excTraceback)
  File "/opt/synapse/lib/python3.9/site-packages/twisted/python/compat.py", line 464, in reraise
    raise exception.with_traceback(traceback)
  File "/opt/synapse/lib/python3.9/site-packages/twisted/enterprise/adbapi.py", line 297, in _runWithConnection
    result = func(conn, *args, **kw)
  File "/opt/synapse/lib/python3.9/site-packages/synapse/storage/database.py", line 738, in inner_func
    return func(db_conn, *args, **kwargs)
  File "/opt/synapse/lib/python3.9/site-packages/synapse/storage/database.py", line 548, in new_transaction
    conn.rollback()
  File "/opt/synapse/lib/python3.9/site-packages/synapse/storage/database.py", line 160, in rollback
    self.conn.rollback()
  File "/opt/synapse/lib/python3.9/site-packages/twisted/enterprise/adbapi.py", line 70, in rollback
    raise ConnectionLost()
twisted.enterprise.adbapi.ConnectionLost

2021-02-27 13:18:55,679 - synapse.http.server - 90 - ERROR - GET-85063- Failed handle request via 'SyncRestServlet': <XForwardedForRequest at 0x7ffb2e2e55e0 method='GET' uri='/_matrix/client/r0/sync?filter=181&set_presence=offline&timeout=0' clientproto='HTTP/1.0' site=8018>
2021-02-27 13:18:55,703 - synapse.access.http.8018 - 299 - INFO - GET-85063- 86.16.94.189 - 8018 - {@Half-Shot:half-shot.uk} Processed request: 0.336sec/-0.000sec (0.009sec, 0.004sec) (0.000sec/0.000sec/0) 55B 500 "GET /_matrix/client/r0/sync?filter=181&set_presence=offline&timeout=0 HTTP/1.0" "Element/1.1.0 (Linux; U; Android 11; Pixel 3a Build/RQ1A.210105.00; Flavour GooglePlay; MatrixAndroidSDK_X 0.0.1)" [0 dbevts]
```

Given that I am 100% certain that my other devices are actively getting new events and hitting the DB successfully on the same process, I am confused as to why the Element android device might be seeing this over and over. It *feels* like the failure was cached, especially due to the quick response time.

### PR DESCRIPTION
**Pull request by ShadowJonathan**
This fixes #9507 by making `ResponseCache` additionally not cache the result when;
- The original function throws an exception.
- One of the conditional callables, upon executing, throws an exception.

Previously, the latter could cause the cache not to be evicted inside the `errback` of the `ObservableDeferred`, because these conditionals would be evaluated (and have an exception thrown) before a `callLater` or immidiate cache eviction could be done, this what caused #9507 (most likely).

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
* [x] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#sign-off)
* [x] Code style is correct (run the [linters](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#code-style))

`Signed-off-by: Jonathan de Jong <jonathan@automatia.nl>`

Fixes #9507

### CODE REVIEW

**Comment by ShadowJonathan** [comment]
(I also split the functions into class-level private methods, because it was getting too crowded with 3 local functions)

File: synapse/util/caches/response_cache.py [hunk 1]
```diff
@@ -80,6 +81,42 @@ def get(self, key: T) -> Optional[defer.Deferred]:
...
+
+    def _errback_remove_cache(self, f: Failure, key: T):
+        self._remove_cache(key)
+        return f
+
+    def _test_cache(self, r: Any, key: T):
+        should_cache = all(
+            self._safe_call(func, r, key=key)
+            for func in self.pending_conditionals.pop(key, [])
+        )
```
Comment (richvdh) [inline synapse/util/caches/response_cache.py:112]: for links: I've argued in #synapse-dev and https://github.com/matrix-org/synapse/pull/9358/files#r585481702 that we should only support one conditional per key, which simplifies this code.

File: changelog.d/9522.bugfix [hunk 2]
```diff
@@ -0,0 +1 @@
+Fix #9507 by hardening ResponseCache against errors in execution of to-be-cached functions.
```
Comment (clokep) [inline changelog.d/9522.bugfix:1]: I think this should be the same message as #9358:

> Added a fix that invalidates cache for empty timed-out sync responses.

Please also rename `changelog.d/9358.misc` -> `changelog.d/9358.bugfix` per https://github.com/matrix-org/synapse/pull/9358#discussion_r585472307

File: synapse/util/caches/response_cache.py (hunk 1 above)
Comment (ShadowJonathan) [reply synapse/util/caches/response_cache.py:112]: I'll add that to the "small changes to `ResponseCache` PR", this one is focused on fixing #9507, i dont wanna drag everything else into here.

File: changelog.d/9522.bugfix (hunk 2 above)
Comment (ShadowJonathan) [reply changelog.d/9522.bugfix:1]: Is it possible to do that without freaking out towncrier?

Comment (clokep) [reply changelog.d/9522.bugfix:1]: towncrier is not a bot, it is just a tool we run during release. So yes, we do this all the time.

File: synapse/util/caches/response_cache.py (hunk 1 above)
Comment (ShadowJonathan) [reply synapse/util/caches/response_cache.py:112]: Created #9525 for this

**Comment by ShadowJonathan** [comment]
Folded into #9739

**Pull request closed without merging**